File: Python/Linked List/singleLL.c
#include <stdio.h>
#include <stdlib.h>

// Define Node structure
typedef struct Node {
    char value;
    struct Node *next;
} Node;

// Define Linked List structure
typedef struct LinkedList {
    Node *head;
} LinkedList;

// Initialize Linked List
void initializeLinkedList(LinkedList *ll) {
    ll->head = NULL;
}

// Add node to the front of the Linked List
void addToFront(LinkedList *ll, char value) {
    Node *newNode = (Node *)malloc(sizeof(Node));
    newNode->value = value;
    newNode->next = NULL;

    if (ll->head == NULL) {
        ll->head = newNode;
    } else {
        newNode->next = ll->head;
        ll->head = newNode;
    }
}

// Add node to the end of the Linked List
void addToEnd(LinkedList *ll, char value) {
    Node *newNode = (Node *)malloc(sizeof(Node));
    newNode->value = value;
    newNode->next = NULL;

    if (ll->head == NULL) {
        ll->head = newNode;
    } else {
        Node *curr = ll->head;
        while (curr->next != NULL) {
            curr = curr->next;
        }
        curr->next = newNode;
    }
}

// Delete the first node of the Linked List
void deleteFront(LinkedList *ll) {
    if (ll->head != NULL) {
        Node *temp = ll->head;
        ll->head = ll->head->next;
        free(temp);
    }
}

// Delete the last node of the Linked List
void deleteEnd(LinkedList *ll) {
    if (ll->head != NULL) {
        if (ll->head->next == NULL) {
            free(ll->head);
            ll->head = NULL;
            return;
        }
        Node *curr = ll->head;
        while (curr->next->next != NULL) {
            curr = curr->next;
        }
        free(curr->next);
        curr->next = NULL;
    }
}

// Display the Linked List
void display(LinkedList *ll) {
    Node *curr = ll->head;
    while (curr != NULL) {
        printf("%c\n", curr->value);
        curr = curr->next;
    }
}

int main() {
    LinkedList myLL;
    initializeLinkedList(&myLL);
    
    addToEnd(&myLL, 'A');
    addToEnd(&myLL, 'B');
    addToEnd(&myLL, 'C');
    addToFront(&myLL, 'c');
    addToFront(&myLL, 'b');
    addToFront(&myLL, 'a');
    deleteFront(&myLL);
    deleteEnd(&myLL);
    display(&myLL);

    return 0;
}

File: Python/Linked List/singleLL.py
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None

class LL:
  def __init__(self):
    self.head = None

  def tambahDepan(self, value):
    newNode = Node(value)

    if self.head is None:
      self.head = newNode

    else:
      newNode.next = self.head
      self.head = newNode

  def tambahBelakang(self, value):
    newNode = Node(value)

    if self.head is None:
      self.head = newNode

    else:
      curr = self.head
      while curr.next is not None:
        curr = curr.next
      curr.next = newNode

  def hapusDepan(self):
    self.head = self.head.next

  def hapusBelakang(self):
    curr = self.head
    while curr.next.next is not None:
      curr = curr.next
    curr.next = None

  def tampil(self):
    curr = self.head
    while curr is not None:
      print(curr.value)
      curr = curr.next

# myLL = LL()
# myLL.tambahBelakang('A')
# myLL.tambahBelakang('B')
# myLL.tambahBelakang('C')
# myLL.tambahDepan('c')
# myLL.tambahDepan('b')
# myLL.tambahDepan('a')
# myLL.hapusDepan()
# myLL.hapusBelakang()
# myLL.tampil()